[flang] Add options -W[no-]unused-dummy-argument and -W[no-]unused-variable

JDPailleux opened the pull request:
Adding support of new warning flags `-W[no-]unused-dummy-argument` and `-W[no-]unused-variable`. 
Used in ECRAD, but not a mandatory flags for the project (https://github.com/ecmwf-ifs/ecrad). 
I think it could be usefull to develop various Fortran projects and avoiding unused dummy arguments / variables.

`-Wunused-dummy-argument` : Warn about unused dummy arguments.
`-Wunused-variable` : Warn whenever a local variable or non-constant static variable is unused aside from its declaration.

tarunprabhu commented:
Does support for these options need to be added in the main driver?

tarunprabhu commented on flang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -932,10 +932,24 @@ static bool parseDiagArgs(CompilerInvocation &res, llvm::opt::ArgList &args,
+      } else if (wArg == "unused-dummy-argument") {
+        res.getFrontendOpts().features.Enable(
+            Fortran::common::LanguageFeature::UnusedDummyArgument);
+      } else if (wArg == "no-unused-dummy-argument") {
+        res.getFrontendOpts().features.Enable(
+            Fortran::common::LanguageFeature::UnusedDummyArgument, false);
+      } else if (wArg == "unused-variable") {
+        res.getFrontendOpts().features.Enable(
+            Fortran::common::LanguageFeature::UnusedVariable);
+      } else if (wArg == "no-unused-variable") {
+        res.getFrontendOpts().features.Enable(
+            Fortran::common::LanguageFeature::UnusedVariable, false);
       } else {
-        const unsigned diagID =
-            diags.getCustomDiagID(clang::DiagnosticsEngine::Error,
-                                  "Only `-Werror` is supported currently.");
+        const unsigned diagID = diags.getCustomDiagID(
+            clang::DiagnosticsEngine::Error,
+            "Only `-Werror`, `-W[no]unused-dummy-argument` "
+            "and `-W[no]unused-variable` are supported "
+            "currently.");
```
This doesn't look like it is scalable. For every warning option that is added, we would have to add a check and change the diagnostic line. Besides, the comment above suggests that this was intended to be a temporary measure. This code predates my involvement with this project. Perhaps those who have been around longer have some insight on how the warning classes were intended to be handled: @banach-space @kiranchandramohan @jeanPerier

banach-space replied:
When we implemented this part, Flang had no mechanism to control diagnostics or warnings, which is why this logic is so basic. I see that this is slowly changing.

> This doesn't look like it is scalable.

Agreed - this is not a sustainable long-term approach. One way forward could be to take inspiration from Clang, for example: [clang/include/clang/Basic/DiagnosticGroups.td#L863](https://github.com/llvm/llvm-project/blob/055872acc28afdd8d29acdbec24f4bd415481d33).

My perspective on changes like this is that we shouldn't block them unless we can propose a concrete alternative. In practice, that means someone needs to design and implement a better approach.

@JDPailleux, are we expecting more patches like this in the near future? It would be helpful to understand the urgency of updating this logic.

JDPailleux replied:
@banach-space  Yes, probably, because in the near future we'd like to add more in order to support `-Wall` in Flang. And I agree that in the long term, adding all the warning flags should be improved.

JDPailleux replied:
So, basically, it would be better to add an entry for each warning flag in the main driver? And therefore remove/rework the use of `OPT_W_Joined` in a future PR? I'll modify this PR to do that ?

banach-space replied:
I've just realised that the link that I shared was wrong, I meant this: https://github.com/llvm/llvm-project/blob/055872acc28afdd8d29acdbec24f4bd415481d33/clang/include/clang/Basic/DiagnosticGroups.td#L863

> it would be better to add an entry for each warning flag in the main driver? 

Adding things in the main driver makes sense if it's going to be re-used by Clang. Is it? IIRC, Clang ties these "warning" flags to various diagnostic groups (hence the link above), as opposed to using https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Driver/Options.td.

But please keep in mind that I've not touched this in ~3yrs and these are just bits from my memory. In practice, we should re-visit the logic in Clang and see whether there's scope for re-use. 
* **If yes**, great, just do it!
* **If not**, then Flang needs something bespoke.

As I no longer actively work in this area, I am unable to make a specific suggestion. But hopefully this provides a good starting point.

JDPailleux replied:
Thanks for the tip, I'll look into it!

JDPailleux commented:
> > > Does support for these options need to be added in the main driver?
> > 
> > 
> > There's no need to add these options warning in the main driver, as `OPT_W_Joined` already takes care of this and will now accept `[no]unused-dummy-argument` and `[no]unused-variable` as argument.
> 
> Do these warning flags appear in `flang --help`?

Yes here : `  -W<warning>             Enable the specified warning` (like in clang)

tarunprabhu commented:
> Yes here : ` -W<warning> Enable the specified warning` (like in clang)

If a user doesn't know of the existence of these warning flags, how would they find it? 

Having said that, it seems that one can't do that with `clang` either. Ideally, we should support a page like [this](https://clang.llvm.org/docs/DiagnosticsReference.html) at least, but that's separate from this PR.

banach-space commented:
> Having said that, it seems that one can't do that with `clang` either. Ideally, we should support a page like [this](https://clang.llvm.org/docs/DiagnosticsReference.html) at least, but that's separate from this PR.

That's likely linked to https://github.com/llvm/llvm-project/blob/055872acc28afdd8d29acdbec24f4bd415481d33/clang/include/clang/Basic/DiagnosticGroups.td#L863 (or similar *.td file)

JDPailleux commented:
> > (or similar *.td file)
> 
> From `clang/docs/CMakeLists.txt`
> 
> ```
> gen_rst_file_from_td(DiagnosticsReference.rst -gen-diag-docs ../include/clang/Basic/Diagnostic.td "${docs_targets}")
> ```
> 
> @JDPailleux, will you be looking into a different way of structuring the warning flags as part of this patch?
> 
> I have requested reviews from others for the non-driver parts of this PR.

@tarunprabhu Yes, it's planned to restructure the warnings flags in this patch by using the Clang Diagnostics or a similar mechanism.

JDPailleux committed:
[flang] Add options -W[no-]unused-dummy-argument and -W[no-]unused-variable

Addition of two new warning options and a diagnostic system for Flang
based on the Clang's diagnostics.

tarunprabhu commented:
Thanks for working on this @JDPailleux. 

This looks like a substantial change including to parts of clang. Is it possible to split this into two. One PR for just the "infrastructure" and another for the flang-specifc warnings. This would make it easier to focus the discussion on the design. Could you also provide a high-level description of the approach that you have taken? 

Thanks.

JDPailleux commented:
> Thanks for working on this @JDPailleux.
> 
> This looks like a substantial change including to parts of clang. Is it possible to split this into two. One PR for just the "infrastructure" and another for the flang-specifc warnings. This would make it easier to focus the discussion on the design. Could you also provide a high-level description of the approach that you have taken?
> 
> Thanks.

Yes of course, I'll split this in 2 PRs.

JDPailleux commented:
Hi, A PR has been created to support a diagnostic system for Flang here: https://github.com/llvm/llvm-project/pull/130593. 
I'm open to discuss and apply modifications or rework if necessary.
